README.rst
cligj
======

.. image:: https://travis-ci.org/mapbox/cligj.svg
   :target: https://travis-ci.org/mapbox/cligj

.. image:: https://coveralls.io/repos/mapbox/cligj/badge.png?branch=master
   :target: https://coveralls.io/r/mapbox/cligj?branch=master

Common arguments and options for GeoJSON processing commands, using Click.

`cligj` is for Python developers who create command line interfaces for geospatial data.
`cligj` allows you to quickly build consistent, well-tested and interoperable CLIs for handling GeoJSON.


Arguments
---------

``files_in_arg``
Multiple files

``files_inout_arg``
Multiple files, last of which is an output file.

``features_in_arg``
GeoJSON Features input which accepts multiple representations of GeoJSON features
and returns the input data as an iterable of GeoJSON Feature-like dictionaries

Options
--------

``verbose_opt``

``quiet_opt``

``format_opt``

JSON formatting options
~~~~~~~~~~~~~~~~~~~~~~~

``indent_opt``

``compact_opt``

Coordinate precision option
~~~~~~~~~~~~~~~~~~~~~~~~~~~
``precision_opt``

Geographic (default), projected, or Mercator switch
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
``projection_geographic_opt``

``projection_projected_opt``

``projection_mercator_opt``

Feature collection or feature sequence switch
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
``sequence_opt``

``use_rs_opt``

GeoJSON output mode option
~~~~~~~~~~~~~~~~~~~~~~~~~~
``geojson_type_collection_opt``

``geojson_type_feature_opt``

``def geojson_type_bbox_opt``

Example
-------

Here's an example of a command that writes out GeoJSON features as a collection
or, optionally, a sequence of individual features. Since most software that
reads and writes GeoJSON expects a text containing a single feature collection,
that's the default, and a LF-delimited sequence of texts containing one GeoJSON
feature each is a feature that is turned on using the ``--sequence`` option.
To write sequences of feature texts that conform to the `GeoJSON Text Sequences
standard <https://tools.ietf.org/html/rfc8142>`__ (and might contain
pretty-printed JSON) with the ASCII Record Separator (0x1e) as a delimiter, use
the ``--rs`` option

.. warning:: Future change warning
   GeoJSON sequences (`--sequence`), not collections (`--no-sequence`), will be
   the default in version 1.0.0.


.. code-block:: python

    import click
    import cligj
    import json

    def process_features(features):
        for feature in features:
            # TODO process feature here
            yield feature

    @click.command()
    @cligj.features_in_arg
    @cligj.sequence_opt
    @cligj.use_rs_opt
    def pass_features(features, sequence, use_rs):
        if sequence:
            for feature in process_features(features):
                if use_rs:
                    click.echo(u'\x1e', nl=False)
                click.echo(json.dumps(feature))
        else:
            click.echo(json.dumps(
                {'type': 'FeatureCollection',
                 'features': list(process_features(features))}))

On the command line, the generated help text explains the usage

.. code-block:: console

    Usage: pass_features [OPTIONS] FEATURES...

    Options:
    --sequence / --no-sequence  Write a LF-delimited sequence of texts
                                containing individual objects or write a single
                                JSON text containing a feature collection object
                                (the default).
    --rs / --no-rs              Use RS (0x1E) as a prefix for individual texts
                                in a sequence as per http://tools.ietf.org/html
                                /draft-ietf-json-text-sequence-13 (default is
                                False).
    --help                      Show this message and exit.

And can be used like this

.. code-block:: console

    $ cat data.geojson
    {'type': 'FeatureCollection', 'features': [{'type': 'Feature', 'id': '1'}, {'type': 'Feature', 'id': '2'}]}

    $ pass_features data.geojson
    {'type': 'FeatureCollection', 'features': [{'type': 'Feature', 'id': '1'}, {'type': 'Feature', 'id': '2'}]}

    $ cat data.geojson | pass_features
    {'type': 'FeatureCollection', 'features': [{'type': 'Feature', 'id': '1'}, {'type': 'Feature', 'id': '2'}]}

    $ cat data.geojson | pass_features --sequence
    {'type': 'Feature', 'id': '1'}
    {'type': 'Feature', 'id': '2'}

    $ cat data.geojson | pass_features --sequence --rs
    ^^{'type': 'Feature', 'id': '1'}
    ^^{'type': 'Feature', 'id': '2'}

In this example, ``^^`` represents 0x1e.


CHANGES.txt
Changes
=======

0.7.2 (2021-05-28)
------------------

- Permit use with click versions 8 and higher.

0.7.1 (2020-11-20)
------------------

- Future deprecation warning added in 0.7.0 has been changed. Cligj version
  1.0.0 will support Python versions >= 3.6.

0.7.0 (2020-10-21)
------------------

- Warn about deprecation of support for Python versions < 3.7 in 1.0.0 (#33).
- Warn about future change in --sequence default when the option is used (#31).

0.6.0 (2020-10-19)
------------------

No changes since 0.6b1.

0.6b1 (2020-10-14)
------------------

Future changes:

- Feature sequences, not collections, will be the default form of output in
  version 1.0 (#20).

Bug fixes:

- Always use `encoding='utf-8'` when opening input GeoJSON (#27).
- Improve docstrings (#22).

0.5.0 (2018-09-26)
------------------

- Version 0.4.0 is not compatible with click 7, but version 0.5.0 is (#23).
- The features_in_arg handler now takes sequences of geometry objects (#14).
- The iter_features function has a new per-feature callback and is ready for
  use in other projects like Fiona (#15).
- The plugins module has been removed (#17).

0.4.0 (2015-12-17)
------------------

- Introduces a click argument, `features_in_arg`, which utilizes a click
  callback to normalize the input of geojson features (#9).
- Release from tagged Travis CI builds (#10).

0.3.0 (2015-08-12)
------------------

- Deprecation of the cligj.plugins module (#6). Please switch to the
  click-plugins module: https://github.com/click-contrib/click-plugins. The
  cligj.plugins module will be removed from cligj at version 1.0.

0.2.0 (2015-05-28)
------------------

- Addition of a pluggable command group class and a corresponding click-style
  decorator (#2, #3).

0.1.0 (2015-01-06)
------------------

- Initial release: a collection of GeoJSON-related command line arguments and
  options for use with Click (#1).


tests/__init__.py
# Do not delete this file.  It makes the tests directory behave like a Python
# module, which is required to manually register and test plugins.

tests/test_cli.py
import os
import os.path
import sys

import click
import pytest

import cligj


def test_files_in(runner):
    @click.command()
    @cligj.files_in_arg
    def cmd(files):
        for f in files:
            click.echo(f)

    result = runner.invoke(cmd, ['1.tif', '2.tif'])
    assert not result.exception
    assert result.output.splitlines() == [
        os.path.join(os.getcwd(), '1.tif'),
        os.path.join(os.getcwd(), '2.tif'),
    ]


def test_files_inout(runner):
    @click.command()
    @cligj.files_inout_arg
    def cmd(files):
        for f in files:
            click.echo(f)

    result = runner.invoke(cmd, ['1.tif', '2.tif'])
    assert not result.exception
    assert result.output.splitlines() == [
        os.path.join(os.getcwd(), '1.tif'),
        os.path.join(os.getcwd(), '2.tif'),
    ]


def test_verbose(runner):
    @click.command()
    @cligj.verbose_opt
    def cmd(verbose):
        click.echo("%s" % verbose)

    result = runner.invoke(cmd, ['-vv'])
    assert not result.exception
    assert result.output.splitlines() == ['2']


def test_quiet(runner):
    @click.command()
    @cligj.quiet_opt
    def cmd(quiet):
        click.echo("%s" % quiet)

    result = runner.invoke(cmd, ['-qq'])
    assert not result.exception
    assert result.output.splitlines() == ['2']


def test_format(runner):
    @click.command()
    @cligj.format_opt
    def cmd(driver):
        click.echo("%s" % driver)

    result = runner.invoke(cmd, ['--driver', 'lol'])
    assert not result.exception
    assert result.output.splitlines() == ['lol']

    result = runner.invoke(cmd, ['--format', 'lol'])
    assert not result.exception
    assert result.output.splitlines() == ['lol']

    result = runner.invoke(cmd, ['-f', 'lol'])
    assert not result.exception
    assert result.output.splitlines() == ['lol']

    result = runner.invoke(cmd)
    assert not result.exception
    assert result.output.splitlines() == ['GTiff']


def test_indent(runner):
    @click.command()
    @cligj.indent_opt
    def cmd(indent):
        click.echo("%s" % indent)

    result = runner.invoke(cmd, ['--indent', '2'])
    assert not result.exception
    assert result.output.splitlines() == ['2']

    result = runner.invoke(cmd)
    assert not result.exception
    assert result.output.splitlines() == ['None']


def test_compact(runner):
    @click.command()
    @cligj.compact_opt
    def cmd(compact):
        click.echo("%s" % compact)

    result = runner.invoke(cmd, ['--compact'])
    assert not result.exception
    assert result.output.splitlines() == ['True']

    result = runner.invoke(cmd, ['--not-compact'])
    assert not result.exception
    assert result.output.splitlines() == ['False']

    result = runner.invoke(cmd)
    assert not result.exception
    assert result.output.splitlines() == ['False']


def test_precision(runner):
    @click.command()
    @cligj.precision_opt
    def cmd(precision):
        click.echo("%s" % precision)

    result = runner.invoke(cmd, ['--precision', '2'])
    assert not result.exception
    assert result.output.splitlines() == ['2']

    result = runner.invoke(cmd)
    assert not result.exception
    assert result.output.splitlines() == ['-1']


def test_projection(runner):
    @click.command()
    @cligj.projection_geographic_opt
    @cligj.projection_projected_opt
    @cligj.projection_mercator_opt
    def cmd(projection):
        click.echo("%s" % projection)

    result = runner.invoke(cmd, ['--geographic'])
    assert not result.exception
    assert result.output.splitlines() == ['geographic']

    result = runner.invoke(cmd, ['--projected'])
    assert not result.exception
    assert result.output.splitlines() == ['projected']

    result = runner.invoke(cmd, ['--mercator'])
    assert not result.exception
    assert result.output.splitlines() == ['mercator']

    result = runner.invoke(cmd)
    assert not result.exception
    assert result.output.splitlines() == ['geographic']


@pytest.mark.filterwarnings("ignore")
@pytest.mark.parametrize(
    ("opt", "val"),
    [
        ("--sequence", True),
        ("--no-sequence", False),
        (None, cligj.__version__.startswith("1.0")),
    ],
)
def test_sequence(runner, opt, val):
    """True becomes the default in 1.0"""
    @click.command()
    @cligj.sequence_opt
    def cmd(sequence):
        click.echo(str(sequence))

    result = runner.invoke(cmd, [opt] if opt is not None else [])
    assert not result.exception
    assert result.output.splitlines() == [str(val)]


@pytest.mark.skipif(sys.version_info < (3,), reason="Requires Python 3")
@pytest.mark.xfail(cligj.__version__.startswith("1.0"), reason="No warning in 1.0")
def test_sequence_warns(runner):
    """Warn about --sequence until 1.0"""
    @click.command()
    @cligj.sequence_opt
    def cmd(sequence):
        click.echo(str(sequence))

    with pytest.warns(FutureWarning):
        result = runner.invoke(cmd, ["--sequence"])


@pytest.mark.filterwarnings("ignore")
@pytest.mark.parametrize(("opt", "val"), [("--rs", True), (None, False)])
def test_sequence_rs(runner, opt, val):
    @click.command()
    @cligj.sequence_opt
    @cligj.use_rs_opt
    def cmd(sequence, use_rs):
        click.echo(str(sequence))
        click.echo(str(use_rs))

    result = runner.invoke(cmd, ["--sequence"] + ([opt] if opt is not None else []))
    assert not result.exception
    assert result.output.splitlines() == ["True", str(val)]


@pytest.mark.parametrize(
    ("opt", "val"),
    [("--collection", "collection"), ("--feature", "feature"), ("--bbox", "bbox")],
)
def test_geojson_type(runner, opt, val):
    @click.command()
    @cligj.geojson_type_collection_opt(True)
    @cligj.geojson_type_feature_opt(False)
    @cligj.geojson_type_bbox_opt(False)
    def cmd(geojson_type):
        click.echo(str(geojson_type))

    result = runner.invoke(cmd, [opt])
    assert not result.exception
    assert result.output.splitlines() == [val]


tests/test_plugins.py
"""Unittests for ``cligj.plugins``."""


import os
from pkg_resources import EntryPoint
from pkg_resources import iter_entry_points
from pkg_resources import working_set

import click

import cligj.plugins


# Create a few CLI commands for testing
@click.command()
@click.argument('arg')
def cmd1(arg):
    """Test command 1"""
    click.echo('passed')

@click.command()
@click.argument('arg')
def cmd2(arg):
    """Test command 2"""
    click.echo('passed')


# Manually register plugins in an entry point and put broken plugins in a
# different entry point.

# The `DistStub()` class gets around an exception that is raised when
# `entry_point.load()` is called.  By default `load()` has `requires=True`
# which calls `dist.requires()` and the `cligj.plugins.group()` decorator
# doesn't allow us to change this.  Because we are manually registering these
# plugins the `dist` attribute is `None` so we can just create a stub that
# always returns an empty list since we don't have any requirements.  A full
# `pkg_resources.Distribution()` instance is not needed because there isn't
# a package installed anywhere.
class DistStub(object):
    def requires(self, *args):
        return []

working_set.by_key['cligj']._ep_map = {
    'cligj.test_plugins': {
        'cmd1': EntryPoint.parse(
            'cmd1=tests.test_plugins:cmd1', dist=DistStub()),
        'cmd2': EntryPoint.parse(
            'cmd2=tests.test_plugins:cmd2', dist=DistStub())
    },
    'cligj.broken_plugins': {
        'before': EntryPoint.parse(
            'before=tests.broken_plugins:before', dist=DistStub()),
        'after': EntryPoint.parse(
            'after=tests.broken_plugins:after', dist=DistStub()),
        'do_not_exist': EntryPoint.parse(
            'do_not_exist=tests.broken_plugins:do_not_exist', dist=DistStub())
    }
}


# Main CLI groups - one with good plugins attached and the other broken
@cligj.plugins.group(plugins=iter_entry_points('cligj.test_plugins'))
def good_cli():
    """Good CLI group."""
    pass


@cligj.plugins.group(plugins=iter_entry_points('cligj.broken_plugins'))
def broken_cli():
    """Broken CLI group."""
    pass


def test_registered():
    # Make sure the plugins are properly registered.  If this test fails it
    # means that some of the for loops in other tests may not be executing.
    assert len([ep for ep in iter_entry_points('cligj.test_plugins')]) > 1
    assert len([ep for ep in iter_entry_points('cligj.broken_plugins')]) > 1


def test_register_and_run(runner):

    result = runner.invoke(good_cli)
    assert result.exit_code is 0

    for ep in iter_entry_points('cligj.test_plugins'):
        cmd_result = runner.invoke(good_cli, [ep.name, 'something'])
        assert cmd_result.exit_code is 0
        assert cmd_result.output.strip() == 'passed'


def test_broken_register_and_run(runner):

    result = runner.invoke(broken_cli)
    assert result.exit_code is 0
    assert u'\U0001F4A9' in result.output or u'\u2020' in result.output

    for ep in iter_entry_points('cligj.broken_plugins'):
        cmd_result = runner.invoke(broken_cli, [ep.name])
        assert cmd_result.exit_code is not 0
        assert 'Traceback' in cmd_result.output


def test_group_chain(runner):

    # Attach a sub-group to a CLI and get execute it without arguments to make
    # sure both the sub-group and all the parent group's commands are present
    @good_cli.group()
    def sub_cli():
        """Sub CLI."""
        pass

    result = runner.invoke(good_cli)
    assert result.exit_code is 0
    assert sub_cli.name in result.output
    for ep in iter_entry_points('cligj.test_plugins'):
        assert ep.name in result.output

    # Same as above but the sub-group has plugins
    @good_cli.group(plugins=iter_entry_points('cligj.test_plugins'))
    def sub_cli_plugins():
        """Sub CLI with plugins."""
        pass

    result = runner.invoke(good_cli, ['sub_cli_plugins'])
    assert result.exit_code is 0
    for ep in iter_entry_points('cligj.test_plugins'):
        assert ep.name in result.output

    # Execute one of the sub-group's commands
    result = runner.invoke(good_cli, ['sub_cli_plugins', 'cmd1', 'something'])
    assert result.exit_code is 0
    assert result.output.strip() == 'passed'


cligj/plugins.py
"""
Common components required to enable setuptools plugins.

In general the components defined here are slightly modified or subclassed
versions of core click components.  This is required in order to insert code
that loads entry points when necessary while still maintaining a simple API
is only slightly different from the click API.  Here's how it works:

When defining a main commandline group:

    >>> import click
    >>> @click.group()
    ... def cli():
    ...    '''A commandline interface.'''
    ...    pass

The `click.group()` decorator turns `cli()` into an instance of `click.Group()`.
Subsequent commands hang off of this group:

    >>> @cli.command()
    ... @click.argument('val')
    ... def printer(val):
    ...    '''Print a value.'''
    ...    click.echo(val)

At this point the entry points, which are just instances of `click.Command()`,
can be added to the main group with:

    >>> from pkg_resources import iter_entry_points
    >>> for ep in iter_entry_points('module.commands'):
    ...    cli.add_command(ep.load())

This works but its not very Pythonic, is vulnerable to typing errors, must be
manually updated if a better method is discovered, and most importantly, if an
entry point throws an exception on completely crashes the group the command is
attached to.

A better time to load the entry points is when the group they will be attached
to is instantiated.  This requires slight modifications to the `click.group()`
decorator and `click.Group()` to let them load entry points as needed.  If the
modified `group()` decorator is used on the same group like this:

    >>> from pkg_resources import iter_entry_points
    >>> import cligj.plugins
    >>> @cligj.plugins.group(plugins=iter_entry_points('module.commands'))
    ... def cli():
    ...    '''A commandline interface.'''
    ...    pass

Now the entry points are loaded before the normal `click.group()` decorator
is called, except it returns a modified `Group()` so if we hang another group
off of `cli()`:

    >>> @cli.group(plugins=iter_entry_points('other_module.commands'))
    ... def subgroup():
    ...    '''A subgroup with more plugins'''
    ...    pass

We can register additional plugins in a sub-group.

Catching broken plugins is done in the modified `group()` which attaches instances
of `BrokenCommand()` to the group instead of instances of `click.Command()`.  The
broken commands have special help messages and override `click.Command.invoke()`
so the user gets a useful error message with a traceback if they attempt to run
the command or use `--help`.
"""


import os
import sys
import traceback
import warnings

import click


warnings.warn(
    "cligj.plugins has been deprecated in favor of click-plugins: "
    "https://github.com/click-contrib/click-plugins",
    FutureWarning, stacklevel=2)


class BrokenCommand(click.Command):

    """
    Rather than completely crash the CLI when a broken plugin is loaded, this
    class provides a modified help message informing the user that the plugin is
    broken and they should contact the owner.  If the user executes the plugin
    or specifies `--help` a traceback is reported showing the exception the
    plugin loader encountered.
    """

    def __init__(self, name):

        """
        Define the special help messages after instantiating `click.Command()`.

        Parameters
        ----------
        name : str
            Name of command.
        """

        click.Command.__init__(self, name)

        util_name = os.path.basename(sys.argv and sys.argv[0] or __file__)

        if os.environ.get('CLIGJ_HONESTLY'):  # pragma no cover
            icon = u'\U0001F4A9'
        else:
            icon = u'\u2020'

        self.help = (
            "\nWarning: entry point could not be loaded. Contact "
            "its author for help.\n\n\b\n"
            + traceback.format_exc())
        self.short_help = (
            icon + " Warning: could not load plugin. See `%s %s --help`."
            % (util_name, self.name))

    def invoke(self, ctx):

        """
        Print the error message instead of doing nothing.

        Parameters
        ----------
        ctx : click.Context
            Required for click.
        """

        click.echo(self.help, color=ctx.color)
        ctx.exit(1)  # Defaults to 0 but we want an error code


class Group(click.Group):

    """
    A subclass of `click.Group()` that returns the modified `group()` decorator
    when `Group.group()` is called.  Used by the modified `group()` decorator.
    So many groups...

    See the main docstring in this file for a full explanation.
    """

    def __init__(self, **kwargs):
        click.Group.__init__(self, **kwargs)

    def group(self, *args, **kwargs):

        """
        Return the modified `group()` rather than `click.group()`.  This
        gives the user an opportunity to assign entire groups of plugins
        to their own subcommand group.

        See the main docstring in this file for a full explanation.
        """

        def decorator(f):
            cmd = group(*args, **kwargs)(f)
            self.add_command(cmd)
            return cmd

        return decorator


def group(plugins=None, **kwargs):

    """
    A special group decorator that behaves exactly like `click.group()` but
    allows for additional plugins to be loaded.

    Example:

        >>> import cligj.plugins
        >>> from pkg_resources import iter_entry_points
        >>> plugins = iter_entry_points('module.entry_points')
        >>> @cligj.plugins.group(plugins=plugins)
        ... def cli():
        ...    '''A CLI aplication'''
        ...    pass

    Plugins that raise an exception on load are caught and converted to an
    instance of `BrokenCommand()`, which has better error handling and prevents
    broken plugins from taking crashing the CLI.

    See the main docstring in this file for a full explanation.

    Parameters
    ----------
    plugins : iter
        An iterable that produces one entry point per iteration.
    kwargs : **kwargs
        Additional arguments for `click.Group()`.
    """

    def decorator(f):

        kwargs.setdefault('cls', Group)
        grp = click.group(**kwargs)(f)

        if plugins is not None:
            for entry_point in plugins:
                try:
                    grp.add_command(entry_point.load())

                except Exception:
                    # Catch this so a busted plugin doesn't take down the CLI.
                    # Handled by registering a dummy command that does nothing
                    # other than explain the error.
                    grp.add_command(BrokenCommand(entry_point.name))
        return grp

    return decorator


cligj/features.py
from itertools import chain
import json
import re

import click


def normalize_feature_inputs(ctx, param, value):
    """Click callback that normalizes feature input values.

    Returns a generator over features from the input value.

    Parameters
    ----------
    ctx: a Click context
    param: the name of the argument or option
    value: object
        The value argument may be one of the following:

        1. A list of paths to files containing GeoJSON feature
           collections or feature sequences.
        2. A list of string-encoded coordinate pairs of the form
           "[lng, lat]", or "lng, lat", or "lng lat".

        If no value is provided, features will be read from stdin.
    """
    for feature_like in value or ('-',):
        try:
            with click.open_file(feature_like) as src:
                for feature in iter_features(iter(src)):
                    yield feature
        except IOError:
            coords = list(coords_from_query(feature_like))
            yield {
                'type': 'Feature',
                'properties': {},
                'geometry': {
                    'type': 'Point',
                    'coordinates': coords}}


def iter_features(geojsonfile, func=None):
    """Extract GeoJSON features from a text file object.

    Given a file-like object containing a single GeoJSON feature
    collection text or a sequence of GeoJSON features, iter_features()
    iterates over lines of the file and yields GeoJSON features.

    Parameters
    ----------
    geojsonfile: a file-like object
        The geojsonfile implements the iterator protocol and yields
        lines of JSON text.
    func: function, optional
        A function that will be applied to each extracted feature. It
        takes a feature object and may return a replacement feature or
        None -- in which case iter_features does not yield.
    """
    func = func or (lambda x: x)
    first_line = next(geojsonfile)

    # Does the geojsonfile contain RS-delimited JSON sequences?
    if first_line.startswith(u'\x1e'):
        text_buffer = first_line.strip(u'\x1e')
        for line in geojsonfile:
            if line.startswith(u'\x1e'):
                if text_buffer:
                    obj = json.loads(text_buffer)
                    if 'coordinates' in obj:
                        obj = to_feature(obj)
                    newfeat = func(obj)
                    if newfeat:
                        yield newfeat
                text_buffer = line.strip(u'\x1e')
            else:
                text_buffer += line
        # complete our parsing with a for-else clause.
        else:
            obj = json.loads(text_buffer)
            if 'coordinates' in obj:
                obj = to_feature(obj)
            newfeat = func(obj)
            if newfeat:
                yield newfeat

    # If not, it may contains LF-delimited GeoJSON objects or a single
    # multi-line pretty-printed GeoJSON object.
    else:
        # Try to parse LF-delimited sequences of features or feature
        # collections produced by, e.g., `jq -c ...`.
        try:
            obj = json.loads(first_line)
            if obj['type'] == 'Feature':
                newfeat = func(obj)
                if newfeat:
                    yield newfeat
                for line in geojsonfile:
                    newfeat = func(json.loads(line))
                    if newfeat:
                        yield newfeat
            elif obj['type'] == 'FeatureCollection':
                for feat in obj['features']:
                    newfeat = func(feat)
                    if newfeat:
                        yield newfeat
            elif 'coordinates' in obj:
                newfeat = func(to_feature(obj))
                if newfeat:
                    yield newfeat
                for line in geojsonfile:
                    newfeat = func(to_feature(json.loads(line)))
                    if newfeat:
                        yield newfeat

        # Indented or pretty-printed GeoJSON features or feature
        # collections will fail out of the try clause above since
        # they'll have no complete JSON object on their first line.
        # To handle these, we slurp in the entire file and parse its
        # text.
        except ValueError:
            text = "".join(chain([first_line], geojsonfile))
            obj = json.loads(text)
            if obj['type'] == 'Feature':
                newfeat = func(obj)
                if newfeat:
                    yield newfeat
            elif obj['type'] == 'FeatureCollection':
                for feat in obj['features']:
                    newfeat = func(feat)
                    if newfeat:
                        yield newfeat
            elif 'coordinates' in obj:
                newfeat = func(to_feature(obj))
                if newfeat:
                    yield newfeat


def to_feature(obj):
    """Takes a feature or a geometry
    returns feature verbatim or
    wraps geom in a feature with empty properties
    """
    if obj['type'] == 'Feature':
        return obj
    elif 'coordinates' in obj:
        return {
            'type': 'Feature',
            'properties': {},
            'geometry': obj}
    else:
        raise ValueError("Object is not a feature or geometry")


def iter_query(query):
    """Accept a filename, stream, or string.
    Returns an iterator over lines of the query."""
    try:
        itr = click.open_file(query).readlines()
    except IOError:
        itr = [query]
    return itr


def coords_from_query(query):
    """Transform a query line into a (lng, lat) pair of coordinates."""
    try:
        coords = json.loads(query)
    except ValueError:
        vals = re.split(r"\,*\s*", query.strip())
        coords = [float(v) for v in vals]
    return tuple(coords[:2])


def normalize_feature_objects(feature_objs):
    """Takes an iterable of GeoJSON-like Feature mappings or
    an iterable of objects with a geo interface and
    normalizes it to the former."""
    for obj in feature_objs:
        if hasattr(obj, "__geo_interface__") and \
           'type' in obj.__geo_interface__.keys() and \
           obj.__geo_interface__['type'] == 'Feature':
            yield obj.__geo_interface__
        elif isinstance(obj, dict) and 'type' in obj and \
                obj['type'] == 'Feature':
            yield obj
        else:
            raise ValueError("Did not recognize object {0}"
                             "as GeoJSON Feature".format(obj))


cligj/__init__.py
"""cligj

A package of arguments, options, and parsers for the Python GeoJSON
ecosystem.
"""

import sys
from warnings import warn

import click

from .features import normalize_feature_inputs

__version__ = "0.7.1dev"

if sys.version_info < (3, 6):
    warn("cligj 1.0.0 will require Python >= 3.6", FutureWarning)


# Multiple input files.
files_in_arg = click.argument(
    'files',
    nargs=-1,
    type=click.Path(resolve_path=True),
    required=True,
    metavar="INPUTS...")


# Multiple files, last of which is an output file.
files_inout_arg = click.argument(
    'files',
    nargs=-1,
    type=click.Path(resolve_path=True),
    required=True,
    metavar="INPUTS... OUTPUT")


# Features from files, command line args, or stdin.
# Returns the input data as an iterable of GeoJSON Feature-like
# dictionaries.
features_in_arg = click.argument(
    'features',
    nargs=-1,
    callback=normalize_feature_inputs,
    metavar="FEATURES...")


# Options.
verbose_opt = click.option(
    '--verbose', '-v',
    count=True,
    help="Increase verbosity.")

quiet_opt = click.option(
    '--quiet', '-q',
    count=True,
    help="Decrease verbosity.")

# Format driver option.
format_opt = click.option(
    '-f', '--format', '--driver', 'driver',
    default='GTiff',
    help="Output format driver")

# JSON formatting options.
indent_opt = click.option(
    '--indent',
    type=int,
    default=None,
    help="Indentation level for JSON output")

compact_opt = click.option(
    '--compact/--not-compact',
    default=False,
    help="Use compact separators (',', ':').")

# Coordinate precision option.
precision_opt = click.option(
    '--precision',
    type=int,
    default=-1,
    help="Decimal precision of coordinates.")

# Geographic (default), projected, or Mercator switch.
projection_geographic_opt = click.option(
    '--geographic',
    'projection',
    flag_value='geographic',
    default=True,
    help="Output in geographic coordinates (the default).")

projection_projected_opt = click.option(
    '--projected',
    'projection',
    flag_value='projected',
    help="Output in dataset's own, projected coordinates.")

projection_mercator_opt = click.option(
    '--mercator',
    'projection',
    flag_value='mercator',
    help="Output in Web Mercator coordinates.")

# Feature collection or feature sequence switch.
sequence_opt = click.option(
    '--sequence/--no-sequence',
    default=False,
    help="Write a LF-delimited sequence of texts containing individual "
    "objects or write a single JSON text containing a feature "
    "collection object (the default).",
    callback=lambda ctx, param, value: warn(
        "Sequences of Features, not FeatureCollections, will be the default in version 1.0.0",
        FutureWarning,
    )
    or value,
)

use_rs_opt = click.option(
    '--rs/--no-rs',
    'use_rs',
    default=False,
    help="Use RS (0x1E) as a prefix for individual texts in a sequence "
         "as per http://tools.ietf.org/html/draft-ietf-json-text-sequence-13 "
         "(default is False).")


def geojson_type_collection_opt(default=False):
    """GeoJSON FeatureCollection output mode"""
    return click.option(
        '--collection',
        'geojson_type',
        flag_value='collection',
        default=default,
        help="Output as GeoJSON feature collection(s).")


def geojson_type_feature_opt(default=False):
    """GeoJSON Feature or Feature sequence output mode"""
    return click.option(
        '--feature',
        'geojson_type',
        flag_value='feature',
        default=default,
        help="Output as GeoJSON feature(s).")


def geojson_type_bbox_opt(default=False):
    """GeoJSON bbox output mode"""
    return click.option(
        '--bbox',
        'geojson_type',
        flag_value='bbox',
        default=default,
        help="Output as GeoJSON bounding box array(s).")
